lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+		errBuf, err := proto.Marshal(&quicpeeringv1a.DialResponse{
+			Status: status.Convert(trail.ToGRPC(toSend)).Proto(),
+		})
+		if err != nil {
+			return
+		}
+		st.SetWriteDeadline(time.Now().Add(10 * time.Second))
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, uint32(len(errBuf)))); err != nil {
+			return
+		}
+		if _, err := st.Write(errBuf); err != nil {
+			return
+		}
+		if err := st.Close(); err != nil {
+			return
+		}
+	}
+
+	st.SetReadDeadline(time.Now().Add(10 * time.Second))
+	var reqLen uint32
+	if err := binary.Read(st, binary.LittleEndian, &reqLen); err != nil {
+		log.DebugContext(c.Context(), "Failed to read request size.", "error", err)
+		return
+	}
+	if reqLen >= quicMaxMessageSize {
+		log.WarnContext(c.Context(), "Received oversized request.", "request_len", reqLen)
+		return
+	}
+	reqBuf := make([]byte, reqLen)
+	if _, err := io.ReadFull(st, reqBuf); err != nil {
+		log.DebugContext(c.Context(), "Failed to read request.", "error", err)
+		return
+	}
+	st.SetReadDeadline(time.Time{})
+
+	req := new(quicpeeringv1a.DialRequest)
+	if err := proto.Unmarshal(reqBuf, req); err != nil {
+		log.WarnContext(c.Context(), "Failed to unmarshal request.", "error", err)
+		return
+	}
+
+	if requestTimestamp := req.GetTimestamp().AsTime(); time.Since(requestTimestamp).Abs() > quicTimestampGraceWindow {
+		log.WarnContext(c.Context(),
+			"Dial request has out of sync timestamp, 0-RTT performance will be impacted.",
+			"request_timestamp", requestTimestamp,
+		)
+		select {
+		case <-c.HandshakeComplete():
+		case <-c.Context().Done():
+			return
+		}
+	}
+
+	// a replayed request is always wrong even after a full handshake, the
+	// replay might've happened before the legitimate request
+	if !s.replayStore.add(req.GetNonce(), time.Now()) {
+		log.ErrorContext(c.Context(), "Request is reusing a nonce, rejecting.", "nonce", req.GetNonce())
+		sendErr(trace.BadParameter("reused or invalid nonce"))
+		return
+	}
+
+	_, clusterName, ok := strings.Cut(req.GetServerId(), ".")
+	if !ok {
+		sendErr(trace.BadParameter("server_id %q is missing cluster information", req.GetServerId()))
+		return
+	}
+
+	nodeConn, err := s.clusterDialer.Dial(clusterName, DialParams{
+		From: &utils.NetAddr{
+			Addr:        req.GetSource().GetAddr(),
+			AddrNetwork: req.GetSource().GetNetwork(),
+		},
+		To: &utils.NetAddr{
+			Addr:        req.GetDestination().GetAddr(),
+			AddrNetwork: req.GetDestination().GetNetwork(),
+		},
+		ServerID: req.GetServerId(),
+		ConnType: types.TunnelType(req.GetConnType()),
+	})
+	if err != nil {
+		sendErr(err)
+		return
+	}
+	defer nodeConn.Close()
+
+	var eg errgroup.Group
+	eg.Go(func() error {
+		defer st.Close()
+		// an empty protobuf message has an empty wire encoding, so by sending a
+		// size of 0 (i.e. four zero bytes) we are sending an empty DialResponse
+		// with an empty Status, which signifies a successful dial
+		if _, err := st.Write(binary.LittleEndian.AppendUint32(nil, 0)); err != nil {
```

Comment:
This is clever and all but I'd rather we just did a send() call with the corresponding "OK" message.

Reply:
I don't want to bring in a fallible operation that allocates a bunch of times, initializes hidden state and is fallible just to send four NULs.

Reply:
You can marshal an empty message once and reuse if you like, but the readability is needlessly suffering here.